### ISSUE TITLE
Allow sandboxed applications to access the kernel keyrings

### ISSUE THREAD

**Issue by complexspaces** [flatpak/flatpak#4281]
## Linux distribution and version
Ubuntu 20.04
## Flatpak version
1.6.5
## Description of the problem
Any `KEYCTL` related syscalls return a permission denied error.

1Password would like to create an official Flatpak for the 1Password for Linux application, but a few of our features are dependent on access to the kernel keyrings. These currently including system authentication and browser integration. 

It's [noted in the source](https://github.com/flatpak/flatpak/blob/686af7d3b81b1315256e79eb5991cede458f3498/common/flatpak-run.c#L3022) that Flatpak would like to allow desktop apps to access these, but is concerned about privacy. It seems like, to me, that the most straightforward way to allow behavior like this is to add a new configuration option that would only disable the specific seccomp filter for specific packages. I do agree though that it could be a privacy risk since its rather hard to create keys that can't be accessed by random processes.

Would the Flatpak team consider adding opt-in support for this?

## Steps to reproduce
1. Install any application to get a container.
2. Copy/paste an application that uses `KEYCTL` syscalls into the `bin` folder of the Flatpak app.
3. Use `flatpak run --command=/bin/bash org.org.name` to get a shell
4. Try running `keyutils` or another small app.
5. Observe that it gets permission denied errors.

You should also be able to see the audit errors with `sudo journalctl --since=today | grep audit`: 
```
May 19 17:55:16 machine-name kernel: audit: type=1326 audit(1621464916.053:3): auid=1000 uid=1000 gid=1000 ses=3 subj=kernel pid=213966 comm="name:inv" exe="/path/to/binary" sig=0 arch=c000003e syscall=250 compat=0 ip=0xfffffffffff code=0x50000
```
In this case, syscall `250` is `keyctl`.

**Comment by mwleeds** [comment]
According to `keyctl(1)`, the key identifiers `-1` and `-2` are for the thread keyring and process keyring respectively whereas other values search the session keyring, user keyring, etc. I think restricting `keyctl()` calls to only thread/process keyrings would mean data couldn't leak from other processes on the system that aren't in the same sandbox. I'm not sure but maybe seccomp can check the values of arguments and filter that way. If so, would that restriction work for your use case?

**Comment by complexspaces** [comment]
Unfortunately, we are currently using a few more then that.

Unconditionally we use:
- Thread keyring

Majority of distros, including Ubuntu 20, Debian, Arch, and Fedora:
- User keyring

RHEL distros such as RHEL 8, CentOS, Ubuntu 18, and Mint:
- Session keyring

We don't have another option on the latter two since the keys created inside them must be visible across process boundaries.

**Comment by complexspaces** [comment]
Hi @mwleeds, 

Just checking in to ask if there has been any more progress on this one between the Flatpak team?

**Comment by swick** [comment]
Without some more information on why and how you use the kernel keyrings it's hard to come up with a solution.

**Comment by complexspaces** [comment]
A few things have changed since I originally opened this issue, but I can expand more. The tl;dr is that [this comment](https://github.com/flatpak/flatpak/issues/4281#issuecomment-844610292) of mine is mostly irrelevant now. We would be more then happy with just access to the thread keyrings inside Flatpak at a minimum.

We use the thread keyring to store a key to allow for secure [biometric authentication](https://support.1password.com/system-authentication-linux/).

Currently, and this is subject to change, we use the user/session keyring to share an encryption key across process boundaries to help secure the IPC socket between the desktop application and the browser extension. However, this has shown to not be a maintainable approach and will likely be removed in favor of something that's easier to adopt. So I don't believe that part has any relevance to this feature request anymore.

**Comment by ruffson** [comment]
This feature would be very valueable for people who are using the Flatpak version of Firefox and want the 1Password desktop application to communicate with the browser plugin. I hope something can be worked out to make this a reality.

**Comment by refi64** [comment]
So I've spent a bit of time looking into this (1password user here!), and unfortunately I think the verdict is that, at the moment, there isn't really a practical way to do this.

First off: we can't just restrict searches to the process/thread keyring. In order to read the value of a key, you use `KEYCTL_READ`, which is given a key ID directly. Thus, although we could restrict *searches* to the process/thread keyring, we cannot restrict *reads*: at the time an seccomp profile would intercept a `KEYCTL_READ` call, it's not clear what keyring said key is from, so it has no way of knowing whether to allow or deny the syscall.

Now, we *could* just assign a new session keyring [like Docker supports](https://github.com/opencontainers/runc/pull/488), which ensure the session keyring is per-Flatpak, however:

> With the patch, each container gets its own session keyring. However, it does work better with user namespaces enabled than without it. With user namespaces each container has its own session keyring _ses and a 'docker exec' joins this session keyring. **Without user namespaces enabled, each container also gets a session keyring on 'docker run' but a 'docker exec' again joins the first created session keyring so that containers share it.**

In other words, from what I can tell, this would break an app's security on `flatpak enter`  if using setuid bwrap.

In theory, allowing access to a private keyring could just be locked behind user namespaces, which is something there *is* precedent for (in particular, Chromium will not run if user namespaces are unavailable). However, this feature is unique in that we can't really handle the failure situation properly. With Chromium, we can detect whether or not the feature is available in userspace and then show an error if it's not. In the keyring case, however, the availability of user namespaces would be impacting the seccomp profile...except user namespace availability is detected by bwrap, but the profile is generated by Flatpak, *before* bwrap is called. Thus, at the time when the seccomp profiles are generated, we have no idea whether or not user namespaces are available. The only other option I can think of is blocking `flatpak enter` in this case, but...I don't feel like that would be an appropriate "solution".

TLDR:

- We can't just use seccomp to allow only keyring calls tied to the process/thread keyring, because the relevant syscalls don't contain information on the keyring the given key ID comes from.
- We can't really follow Docker and give Flatpak apps their own private session keyrings, because this would break security with `flatpak enter`, and we can't know ahead of time if it should be allowed at all until `bwrap` is run.

Now...I do have to ask, @complexspaces, what *exactly* is the thread keyring used for? I understand it's used for biometrics, but it's not clear to me why storing the data in the thread keyring is useful in the first place. I noticed the linked article mentions:

> The secret is not saved to disk and can’t be read by unprivileged processes. It stays in the memory while 1Password is running (including when it runs in the system tray) and is lost when you quit 1Password.

Is this what gets stored in the kernel keyring?

**Comment by complexspaces** [comment]
Hi, sorry for the overdue response here. It fell off my radar for a while unfortunately :( 

Firstly, thank you for the extremely thorough investigation.  _Keyctl-unmask:_ was a good read-through as well :)

To answer your question, yes, that is what gets stored in the thread keyring. We (1Password) have a "locked means locked" policy on all platforms that we can reasonably apply it to, which includes 1Password for Linux. This means that 1Password's process memory should contain no secrets when the app is locked (baring things like garbage collectors and OS allocation reuse in relevant applications, of course).

For a little bit more background: System authentication (also known as biometrics) is powered by vastly different technologies on the different OSes. The substitute unlock key involved is handed off to the OS for secure storage regardless of the application's current lock state. Since there's nothing like macOS's secure enclave on Linux, we determined that the kernel keyring was the best location for a system unlock key on Linux that met our requirements.

**Comment by swick** [comment]
> This means that 1Password's process memory should contain no secrets when the app is locked (baring things like garbage collectors and OS allocation reuse in relevant applications, of course).

Having the secrets in the kernel mappings is fine though? What kind of threat model are you working with?

**Comment by complexspaces** [comment]
> Having the secrets in the kernel mappings is fine though? What kind of threat model are you working with?

On our supported platforms, we don't try to defend against the superuser (`root`, `Administrator` on Windows, etc). If something can access a kernel mapping solely for our process, there isn't much we can do. Keeping the key out of process also makes it easier to reason about the abstraction across all the other OSes the same code is supporting, even if it doesn't give an obvious or noticeable improvement in security for Linux specifically.

**Comment by swick** [comment]
> On our supported platforms, we don't try to defend against the superuser (root, Administrator on Windows, etc).

What are you trying to defend against then?

> Keeping the key out of process also makes it easier to reason about the abstraction across all the other OSes the same code is supporting, even if it doesn't give an obvious or noticeable improvement in security for Linux specifically.

That's not a good reason for supporting kernel keyrings in flatpak. I want to see a real threat model where the kernel keyring provides benefits.

**Comment by refi64** [comment]
Hmm that could definitely work, the primary way you run commands inside of a systemd-nspawn container is similar (it connects to the container's D-Bus instance and uses that to tell the systemd instance inside to run commands). It would also match nicely with already-existing options like `-d`, which I'd imagine you'd often be using for development purposes anyway.

**Comment by oengelha** [comment]
Is there any news on this front?

**Comment by tunix** [comment]
@complexspaces - I wonder if there are any updates on this? Do browsers need any implementation regarding this since I assume the browser extension communicates through the browser process?

For the record, I'm a long time 1Password user and I'd really love to use 1Password with all its features available on Linux. I use bluefin (Fedora Silverblue) as my daily driver and unfortunetely I had to layer 1Password to be able to use its system auth & SSH Agent features. For the browser, I use Brave with flatpak. I really look forward to a 1Password flatpak release that ticks all the boxes on Linux.

**Comment by complexspaces** [comment]
@tunix I think we should keep this post on-topic and not product specific, but to answer your question real quick: 1Password for Linux has not used the kernel keyrings for facilitating secure IPC for a few years now. Its instead using install-time UNIX groups and `setgid` (with no extra permissions).

With that said, we'd still be interested in this feature because the kernel keyrings are used for system authentication in the same way as when I opened this issue in 2021.

**Comment by elbandi** [comment]
We install Goland IDE to our notebook. We develop an app which use keyring, but cannot run/debug in IDE, because this sandbox. 

We want to debug the app, so have to run the app from the editor.

**Comment by aristeu** [comment]
As far as use cases go, my example here: flatpak browser with kerberos ticket: kerberos started using the kernel keyring for storing credentials and a flatpak packaged browser can't use it. In my case, I think I can work around that if I figure the way to make kerberos go back to using /tmp files.
